packages/core/src/application.ts

```diff
@@ -174,20 +201,44 @@ export class Application extends Context implements LifeCycleObserver {
     return this.get<T>(key);
   }
 
+  private assertValidStates(op: string, ...states: string[]) {
+    assert(
+      states.includes(this._state),
+      `Cannot ${op} the application as it is ${this._state}. Valid states are ${states}.`,
+    );
+  }
+
   /**
-   * Start the application, and all of its registered observers.
+   * Start the application, and all of its registered observers. The application
+   * state is checked to ensure the integrity of `start`.
+   *
+   * If the application is already started, no operation is performed.
    */
   public async start(): Promise<void> {
+    this.assertValidStates('start', 'created', 'stopped', 'started');
```

Comment:
Should we define named constants for these magic strings? Preferably as a string enum, see https://www.typescriptlang.org/docs/handbook/enums.html#string-enums.

Reply:
I'm considering so. On the other hand, do we allow other states, such as `booting` and `booted`?